### PR TITLE
Add escape_html_once and escape_html_once_as_safe_html

### PR DESCRIPTION
**Pull request by pushrax**
This is a port of ActiveSupport's `ERB::Util.html_escape_once` method, which uses this regex to match characters that need encoding, while ignoring existing complete HTML entities.

```rb
HTML_ESCAPE_ONCE_REGEXP = /["><']|&(?!([a-zA-Z]+|(#\d+)|(#[xX][\dA-Fa-f]+));)/
```
([source](https://github.com/rails/rails/blob/4acddae05e65ec62114219d0595d78bd13ed889d/activesupport/lib/active_support/core_ext/string/output_safety.rb#L12)).

This PR is more or less a direct port, also based off code in `unescape_ent` from ext/escape_utils/houdini_html_u.c.

cc @rafaelfranca @maximevaillancourt

### CODE REVIEW

**Comment by pushrax** [comment]
There are new failing tests on Ruby 2.5 and 2.6 that are unrelated to the feature in this PR. Can we merge with failing tests and fix in a separate PR?

**Comment by rafaelfranca** [comment]
I have a branch fixing the tests. I can merge tomorrow and we can rebase on top of it.

**Comment by rafaelfranca** [comment]
The code looks good to me. Can you add a benchmark for it and compare with Active Support's implementation so we know the numbers?

**Comment by pushrax** [comment]
~10x faster than ActiveSupport's regex approach.

```
bx ruby benchmark/html_escape.rb
Escaping 154027 bytes of html from https://en.wikipedia.org/wiki/Succession_to_the_British_throne
Warming up --------------------------------------
Rack::Utils.escape_html
                        10.000  i/100ms
Haml::Helpers.html_escape
                        90.000  i/100ms
ERB::Util.html_escape
                        92.000  i/100ms
ERB::Util.html_escape_once
                        16.000  i/100ms
      CGI.escapeHTML    92.000  i/100ms
         String#gsub    16.000  i/100ms
fast_xs_extra#fast_xs_html
                       154.000  i/100ms
EscapeUtils.escape_html
                       142.000  i/100ms
EscapeUtils.escape_html_once
                       138.000  i/100ms
Calculating -------------------------------------
Rack::Utils.escape_html
                        111.331  (± 1.8%) i/s -    560.000  in   5.032251s
Haml::Helpers.html_escape
                        908.935  (± 4.4%) i/s -      4.590k in   5.060070s
ERB::Util.html_escape
                        913.557  (± 3.7%) i/s -      4.600k in   5.042815s
ERB::Util.html_escape_once
                        158.746  (± 4.4%) i/s -    800.000  in   5.048933s
      CGI.escapeHTML    919.482  (± 3.2%) i/s -      4.600k in   5.007776s
         String#gsub    158.884  (± 3.8%) i/s -    800.000  in   5.041979s
fast_xs_extra#fast_xs_html
                          1.545k (± 3.4%) i/s -      7.854k in   5.089957s
EscapeUtils.escape_html
                          1.428k (± 5.0%) i/s -      7.242k in   5.085393s
EscapeUtils.escape_html_once
                          1.395k (± 3.9%) i/s -      7.038k in   5.054010s

Comparison:
fast_xs_extra#fast_xs_html:     1545.1 i/s
EscapeUtils.escape_html:     1428.1 i/s - same-ish: difference falls within error
EscapeUtils.escape_html_once:     1394.8 i/s - 1.11x  slower
      CGI.escapeHTML:      919.5 i/s - 1.68x  slower
ERB::Util.html_escape:      913.6 i/s - 1.69x  slower
Haml::Helpers.html_escape:      908.9 i/s - 1.70x  slower
         String#gsub:      158.9 i/s - 9.72x  slower
ERB::Util.html_escape_once:      158.7 i/s - 9.73x  slower
Rack::Utils.escape_html:      111.3 i/s - 13.88x  slower
```

Also confirmed this patch doesn't meaningfully impact performance of `html_escape` when `escape_once` is false. The difference is well within variance.

File: ext/escape_utils/houdini_html_e.c [hunk 1]
```diff
@@ -44,15 +44,47 @@ static const char *HTML_ESCAPES[] = {
         "&gt;"
 };
 
+static int
+is_entity(const uint8_t *src, size_t size)
+{
+	size_t i = 0;
+
+	if (size == 0 || src[0] != '&')
+		return 0;
```
[the remark this answers is not in the document]
Comment (pushrax) [reply ext/escape_utils/houdini_html_e.c]: Ah yeah I didn't even think about this, I guess stdbool is included somewhere.

**Comment by pushrax** [comment]
> Although I didn't see anything that targets the first part of the regex `/["><']/`.

The first part of the regex is performed by the rest of the code in houdini_escape_html0

**Comment by DerekStride** [comment]
> The first part of the regex is performed by the rest of the code in houdini_escape_html0

I thought as much but just wanted to highlight it incase it was forgotten.

**Pull request closed without merging**